packages/apollo/src/commands/service/check.ts

```diff
@@ -296,79 +305,129 @@ export default class ServiceCheck extends ProjectCommand {
                    * name: implementing service name inside of the graph
                    * sha: git commit hash/docker id. placeholder for now
                    */
-                  task.output = "Creating composed schema against the graph";
                   const {
                     errors,
                     warnings,
                     compositionValidationDetails
                   } = await project.engine.checkPartialSchema({
                     id: configName,
                     graphVariant: tag,
-                    implementingServiceName: flags.serviceName || info.name,
+                    implementingServiceName: serviceName,
                     partialSchema: {
                       sdl: info.sdl
                     }
                   });
 
                   // FIXME: reformat to match other check results
-
                   taskOutput.federation = {
                     errors,
                     warnings
                   };
 
                   if (compositionValidationDetails) {
                     schemaHash = compositionValidationDetails.schemaHash;
-                  } else {
-                    // FIXME: We should provide some better error handling at the GraphQL layer for this
-                    throw new Error(`Federated service could not be composed due to the following errors:
-  ${errors && errors.map(err => (err && err.message) || "").join("\n")}
-                  `);
                   }
-                } else {
-                  schema = await project.resolveSchema({ tag: config.tag });
 
-                  await gitInfo(this.log);
+                  if (errors.length > 0) {
+                    const decodedErrors = errors.filter(Boolean).map(error => {
+                      // We `filter` by `Boolean` above; this will never happen and is just here to appease
+                      // TypeScript.
+                      if (!error) {
+                        return {};
+                      }
+
+                      const match = error.message.match(
```

Comment:
Maybe we should be able to request these capture groups in the shape we want from the backend, rather than parse the error message. I think this needs to be implemented on our end at the composition level in order for them to do that. This would be a great feature request to bubble up through infra into federation / composition!

Reply:
Totally right! @JakeDawkins brought that up too. I'll make a ticket so I don't forget.

Reply:
https://apollographql.atlassian.net/browse/AP-521

Reply:
🤗